---
title: "MiCM: 01 - Alignment"
author: "Ariel Madrigal Aguirre"
date: "2023-10-25"
output:
  html_notebook:
    df_print: paged
    code_folding: show
    toc: yes
    toc_float: 
      collapsed: false
      smooth_scroll: false
---

```{r,echo=FALSE}
htmltools::img(src = knitr::image_uri("images/micm_color_logo.png"), 
               alt = 'logo', 
               style = 'position:absolute; top:30px; right:0; padding:10px; max-width:50%;')
```

This workshop will guide you through the basics of ChIP-seq analysis with hands-on exercises. You will learn how to process ChIP-seq data: perform read alignment, peak calling, quality control, visualization through the genome browser, motif finding and gene set enrichment analysis.

In this tutorial, we are going to focus on performing read alignment. 

The dataset that we will be using today is ChIP-seq for [HIF1a](https://www.genecards.org/cgi-bin/carddisp.pl?gene=HIF1A) and [HIF2a](https://www.genecards.org/cgi-bin/carddisp.pl?gene=EPAS1). HIF1a and HIF2a are the two principal isoforms of the Hypoxia-inducible factor (HIF), which is the major transcriptional regulator of cellular responses to hypoxia. This dataset was generated by Smythies *et al* (2018), and you can find the paper [here](https://doi.org/10.15252/embr.201846401). Because the original data is very big, we are going to work with a subset of the original reads from the study (reads that map to a segment of the chromosome 6).

**Note**: All the code in this tutorial will be performed in the terminal. 

Let's move to the scripts directory (or stay here if already there):
```{bash, eval=F}
cd chip_seq/scripts
```

# 01 - Mapping with bowtie2

You can check the help on how to use bowtie2 by typing:
```{bash}
bowtie2 --help 
```
You can also access the bowtie2 manual in [here](https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml).

The data that we have is paired end data, so we have 2 .fastq files for one sample. They look like this:
```{bash}
ls ../data/fastq
```

Bowtie2 requires a genome index (compressed files that store the reference sequence). For this time, we've already generated the index and can be found here:
```{bash}
ls ../data/reference/GRCh38_chr6
```
This index was created only using the human chromosome 6. You can create your own index or download pre-built ones from the bowtie2 page.

Let's create a directory where to store our alignment
```{bash}
mkdir -p ../results
```
Now we are ready to run bowtie2!
```{bash}
bowtie2 --very-sensitive-local -x ../data/reference/GRCh38_chr6/hg38_chr6 -1 ../data/fastq/HIF1a_Rep1_R1.fastq.gz -2 ../data/fastq/HIF1a_Rep1_R2.fastq.gz -S ../results/HIF1a_Rep1.sam
```
Let's take a look at the parameters that we've used for running bowtie2:

* --very-sensitive-local: Bowtie2 comes with some useful combinations of parameters packaged into shorter "preset" parameters. The preset options that come with Bowtie 2 are designed to cover a wide area of the speed/sensitivity/accuracy trade-off space. "Fast" presets are generally faster but less sensitive, while the "sensitive" presets are generally slower but more sensitive and accurate. The local part means that we are doing 'local' alignment (remember that the default in Bowtie2 is the end-to-end alignment). 
* --x: path to the index
* -1: path to fastq1 
* -2: path to fastq2
* -S: path to the output SAM file

# 02 - Filtering SAM/BAM

For the following steps, we will be using samtools. The manual can be found [here](http://www.htslib.org/doc/samtools.html). Samtools is a compendium of various tools:
```{bash}
samtools --help
```
We can count the number of reads that are in our SAM
```{bash}
samtools view -c ../results/HIF1a_Rep1.sam
```
* -c: Instead of printing the alignments, only count them and print the total number

We need to convert the SAM file into a BAM file. In this same step, we will perform some filtering of the alignments:
```{bash}
samtools view -h -F 1804 -q 30 -b ../results/HIF1a_Rep1.sam -o ../results/HIF1a_Rep1.bam
```
* -h: include header in the output
* -F: Exclude alignments with a given FLAG field. The 1804 FLAG excludes the following reads: read unmapped, mate unmapped, not primary alignment, read quality low, PCR / optical duplicate. You can check the meaning of the SAM FLAGS in [here](https://broadinstitute.github.io/picard/explain-flags.html). More information about samtools flags can be found [here](https://samtools.github.io/hts-specs/SAMv1.pdf).
* -q: Skip alignments with MAPQ smaller than the given value. The MAPQ is a non-negative integer Q = -10 log10 p, where p is an estimate of the probability that the alignment does not correspond to the read's true point of origin. Usually, a good value is between 10-30. 
* -b: Output in the BAM format.
* -o: Output to file

We can also remove reads that map blacklisted regions by [ENCODE](https://github.com/Boyle-Lab/Blacklist) using bedtools. 

```{bash}
bedtools intersect -abam ../results/HIF1a_Rep1.bam -b ../data/blacklist/hg38.blacklist.bed.gz -v > ../results/HIF1a_Rep1.rmblacklist.bam
```

Finally, the next steps deal with marking and removing duplicates. Removing duplicates is usually a good idea, since most likely duplicate reads are originated from PCR, and don't represent 'true' reads. 

```{bash}
# collate: group by name
samtools collate ../results/HIF1a_Rep1.rmblacklist.bam -o ../results/HIF1a_Rep1.collate.bam
# fills in mate coordinates and insert size fields
samtools fixmate -m ../results/HIF1a_Rep1.collate.bam ../results/HIF1a_Rep1.fixmate.bam
# sort bam
samtools sort ../results/HIF1a_Rep1.fixmate.bam -o ../results/HIF1a_Rep1.sorted.bam
# mark and remove duplicates
samtools markdup ../results/HIF1a_Rep1.sorted.bam ../results/HIF1a_Rep1.sorted.markdup.bam # with this command we don't remove duplicates
samtools markdup -r ../results/HIF1a_Rep1.sorted.bam ../results/HIF1a_Rep1.sorted.rmdup.bam
# create index
samtools index ../results/HIF1a_Rep1.sorted.rmdup.bam
```

We can compare the initial number of reads in the fastq file versus the number of reads and fragments after mapping and filtering.
```{bash}
## Function to count the number of reads in a fastq.gz file
function nreadsfastq(){
    zcat ${1} | echo $((`wc -l`/4))
}
nreadsfastq ../data/fastq/HIF1a_Rep1_R1.fastq.gz
nreadsfastq ../data/fastq/HIF1a_Rep1_R2.fastq.gz
```
```{bash}
samtools flagstat ../results/HIF1a_Rep1.sorted.markdup.bam
```
```{bash}
samtools flagstat ../results/HIF1a_Rep1.sorted.rmdup.bam
```

And that's it! Now you are familiar with alignment using bowtie2 and SAM/BAM processing.

# 03 - Exercises

* 1. Perform the alignment steps for the HIF2a data. The input fastq files can be found in the data folder:
```{bash}
ls ../data/fastq/
```

* 2. How would you filter for reads that are unmapped and have a mate unmapped? Give it a try using the unfiltered .sam output file from HIF1a. Remember that you can check the SAM FLAGS in [here](https://broadinstitute.github.io/picard/explain-flags.html).

* 3. How would you run bowtie2 using the end-to-end mode? How do the results compare to the local mode?

 
```{r}
sessionInfo()
```
