---
title: "MiCM: visualization"
author: "Ariel Madrigal Aguirre"
date: "`r Sys.Date()`"
output:
  html_notebook:
    df_print: paged
    code_folding: show
    toc: yes
    toc_float: 
      collapsed: false
      smooth_scroll: false
---

```{r,echo=FALSE}
htmltools::img(src = knitr::image_uri("images/micm_color_logo.png"), 
               alt = 'logo', 
               style = 'position:absolute; top:30px; right:0; padding:10px; max-width:50%;')
```

This workshop will guide you through the basics of ChIP-seq analysis with hands-on exercises. You will learn how to process ChIP-seq data: perform read alignment, peak calling, quality control, visualization through the genome browser, motif finding and gene set enrichment analysis.

**Note**: All the code in this tutorial will be performed in the terminal. 

# UCSC genome browser

## Upload of peak files

One of the most common useful ways to visualize our data is through the [UCSC genome browser](https://genome.ucsc.edu/index.html). We can visualize the peaks that we've generated with macs and see directly the regions of the genome where our protein of interest is binding. 

Let's move to the scripts directory (or stay here if already there):
```{bash, eval=F}
cd chip_seq/scripts
```

To do this, we need to upload the .bed files that we got from macs2
```{bash}
head ../results/macs2/HIF1a_rep1_peaks.bed
```
In the genome browser, visualize the coordinates of a given peak in your .bed file. 

Remember that in the UCSC genome browser you have to specify the positions in this format: chr:start-end. 

## Generation of bigwig 

One of the most common/useful formats to visualize the data is the bigwig format. To generate this, we are going to use [deeptools](https://deeptools.readthedocs.io/en/develop/).

We are going to convert our .BAM file ( from our alignment step) into a .bigwig file

```{bash}
ls ../results/HIF1a_Rep1.sorted.rmdup.bam
```
To convert, we are going to use the bamCoverage command
```{bash}
bamCoverage --help
```
```{bash}
bamCoverage --bam ../results/HIF1a_Rep1.sorted.rmdup.bam -o ../results/HIF1a_Rep1.bigwig --normalizeUsing BPM --extendReads
```
Let's see the parameters of the call:

* --bam: path of the input .bam file.
* -o: path of the output bigwig file. 
* --normalizeUsing: Use one of the entered methods to normalize the number of reads per bin. By default, no normalization is performed. BPM = Bins Per Million mapped reads, same as TPM in RNA-seq. 
* --extendReads: In the paired-end mode ( as our case), it will extend the reads to the mean of the fragment size of all mate reads. 

Now, we should have our .bigwig file:
```{bash}
ls ../results/HIF1a_Rep1.bigwig
```
UCSC doesn't allow to 'upload' directly your .bigwig file. Instead, you have to store them in a web-accessible http, https, or ftp location. One of the options to store .bigwig files is [Cyverse](https://cyverse.org/discovery-environment). 

Because data upload to servers is time-consuming, for this step, we are going to pretend we've already uploaded to data to Cyverse.

Now, open your UCSC genome browser and click in the Custom Tracks button. Then, you will add the following custom track:
```{}
track type=bigWig name="HIF1a_rep1" description="HIF1a bigwig" bigDataUrl=https://data.cyverse.org/dav-anon/iplant/home/arielmadr23/27_HIF_Smythies_2019/myHub/hg38/RCC4_Normoxia_HIF1a_PM14_Rep1.bw
```
You can read more about how to upload data to the UCSC genome browser in here:

* [create custom tracks](https://genome.ucsc.edu/goldenPath/help/customTrack.html)
* [create Track Hubs](https://genome.ucsc.edu/goldenpath/help/hgTrackHubHelp.html) 

```{r}
sessionInfo()
```
