---
title: "MiCM: 04 - visualization"
author: "Ariel Madrigal Aguirre"
date: "2023-10-25"
output:
  html_notebook:
    df_print: paged
    code_folding: show
    toc: yes
    toc_float: 
      collapsed: false
      smooth_scroll: false
---

```{r,echo=FALSE}
htmltools::img(src = knitr::image_uri("images/micm_color_logo.png"), 
               alt = 'logo', 
               style = 'position:absolute; top:30px; right:0; padding:10px; max-width:50%;')
```

This workshop will guide you through the basics of ChIP-seq analysis with hands-on exercises. You will learn how to process ChIP-seq data: perform read alignment, peak calling, quality control, visualization through the genome browser, motif finding and gene set enrichment analysis.

**Note**: All the code in this tutorial will be performed in the terminal. 

# 01 - UCSC genome browser

## Upload of peak files

One of the most common useful ways to visualize our data is through the [UCSC genome browser](https://genome.ucsc.edu/index.html). We can visualize the peaks that we've generated with macs and see directly the regions of the genome where our protein of interest is binding. 

Let's move to the scripts directory (or stay here if already there):
```{bash, eval=F}
cd chip_seq/scripts
```

To do this, we need to upload the .bed files that we got from macs2
```{bash}
head ../results/macs2/HIF1a_rep1_peaks.bed
```
In the genome browser, visualize the coordinates of a given peak in your .bed file. 

Remember that in the UCSC genome browser you have to specify the positions in this format: chr:start-end. 

## Generation of bigwig 

One of the most common/useful formats to visualize the data is the bigwig format. To generate this, we are going to use [deeptools](https://deeptools.readthedocs.io/en/develop/).deepTools is a suite of python tools particularly developed for the efficient analysis of high-throughput sequencing data, such as ChIP-seq, RNA-seq or MNase-seq. It offers a lot of tools for processing BAM and bigwig files.


We are going to convert our .BAM file ( from our alignment step) into a .bigwig file

```{bash}
ls ../results/HIF1a_Rep1.sorted.rmdup.bam
```
To convert, we are going to use the bamCoverage command
```{bash}
bamCoverage --help
```
```{bash}
bamCoverage --bam ../results/HIF1a_Rep1.sorted.rmdup.bam -o ../results/HIF1a_Rep1.bigwig --normalizeUsing BPM --extendReads
```
Let's see the parameters of the call:

* --bam: path of the input .bam file.
* -o: path of the output bigwig file. 
* --normalizeUsing: Use one of the entered methods to normalize the number of reads per bin. By default, no normalization is performed. BPM = Bins Per Million mapped reads, same as TPM in RNA-seq. 
* --extendReads: In the paired-end mode ( as our case), it will extend the reads to the mean of the fragment size of all mate reads. 

Now, we should have our .bigwig file:
```{bash}
ls ../results/HIF1a_Rep1.bigwig
```
UCSC doesn't allow to 'upload' directly your .bigwig file. Instead, you have to store them in a web-accessible http, https, or ftp location. One of the options to store .bigwig files is [Cyverse](https://cyverse.org/discovery-environment). 

Because data upload to servers is time-consuming, for this step, we are going to pretend we've already uploaded to data to Cyverse.

Now, open your UCSC genome browser and click in the Custom Tracks button. Then, you will add the following custom track:
```{}
track type=bigWig name="HIF1a_rep1" description="HIF1a bigwig" bigDataUrl=https://data.cyverse.org/dav-anon/iplant/home/arielmadr23/27_HIF_Smythies_2019/myHub/hg38/RCC4_Normoxia_HIF1a_PM14_Rep1.bw
```
You can read more about how to upload data to the UCSC genome browser in here:

* [Create custom tracks](https://genome.ucsc.edu/goldenPath/help/customTrack.html)
* [Create Track Hubs](https://genome.ucsc.edu/goldenpath/help/hgTrackHubHelp.html) 

# 02 - Deeptools

Deeptools also offers different tools for visualizing ChIP-seq data, including the generation of Heatmaps and different summary plots. For now, let's try the [plotHeatmap function](https://deeptools.readthedocs.io/en/develop/content/tools/plotHeatmap.html). 

```{bash}
mkdir -p ../results/deeptools
```
The first step is to compute the signal using computeMatrix. computeMatrix has two modes: 
* scale regions: In the scale-regions mode, all regions in the BED file are stretched or shrunken to the length (in bases) indicated by the user. 
* reference-point: Reference-point refers to a position within a BED region (e.g., the starting point). In this mode, only those genomicpositions before (upstream) and/or after (downstream) of the reference point will be plotted. You can read more about them [here](https://deeptools.readthedocs.io/en/develop/content/tools/computeMatrix.html). 

```{bash}
computeMatrix reference-point -S ../results/HIF1a_Rep1.bigwig \
			  --referencePoint center \
			  -R ../results/macs2/HIF1a_rep1_summits.bed \
			  --beforeRegionStartLength 3000 \
			  --afterRegionStartLength 3000 \
			  --skipZeros \
			  -o ../results/deeptools/matrix.mat.gz
```
* -S: bigWig file(s) containing the scores to be plotted. Multiple files should be separated by spaced.
* --referencePointThe: reference point for the plotting could be either the region start (TSS), the region end (TES) or the center of the region.
* -R: File name or names, in BED or GTF format, containing the regions to plot. If multiple bed files are given, each one is considered a group that can be plotted separately.
* --beforeRegionStartLength: Distance upstream of the reference-point selected. (Default: 500)
* --afterRegionStartLength: Distance downstream of the reference-point selected. (Default: 1500)
* skipZeros: Whether regions with only scores of zero should be included or not.
* -o: File name to save. 

```{bash}
plotHeatmap -m ../results/deeptools/matrix.mat.gz \
			--heatmapHeight 5  \
			--refPointLabel 'Peak summit' \
			--regionsLabel 'HIF1a peaks' \
			--plotTitle 'Peak summits (HIF1a)' \
			-out ../results/deeptools/HIF1a_heatmap.png
```

![](../results/deeptools/HIF1a_heatmap.png)

* -m: Matrix file from the computeMatrix tool. 
* -out: File name to save the image to. 
* --heatmapHeight: Plot height in cm.
* --refPointLabel: Label shown in the plot for the reference-point. 
* --regionsLabel: Labels for the regions plotted in the heatmap. If more than one region is being plotted, a list of labels separated by spaces is required.
* --plotTitle: Title of the plot, to be printed on top of the generated image

# 03 - Exercises

* 1. Include the HIF2a data in the Heatmap from Deeptools. This heatmap should show the  coverage for both HIF1a and HIF2a peaks. 


```{r}
sessionInfo()
```
